File: clients/python/src/graphdb_client/aio/resources/edges.py
from __future__ import annotations

from typing import Any, AsyncIterator, Mapping, Sequence

from ...models import Edge
from ..transport import AsyncTransport


class AsyncEdgesResource:
    def __init__(self, transport: AsyncTransport) -> None:
        self._t = transport

    async def create(
        self,
        from_node_id: int,
        to_node_id: int,
        edge_type: str,
        *,
        properties: Mapping[str, Any] | None = None,
        weight: float = 0.0,
    ) -> Edge:
        res = await self._t.request("POST", "/edges", json={
            "from_node_id": from_node_id,
            "to_node_id": to_node_id,
            "type": edge_type,
            "properties": dict(properties or {}),
            "weight": weight,
        })
        return Edge.from_dict(res.data)

    async def get(self, edge_id: int) -> Edge:
        res = await self._t.request("GET", f"/edges/{edge_id}")
        return Edge.from_dict(res.data)

    async def update(
        self,
        edge_id: int,
        properties: Mapping[str, Any] | None = None,
        *,
        weight: float | None = None,
    ) -> Edge:
        body: dict[str, Any] = {}
        if properties is not None:
            body["properties"] = dict(properties)
        if weight is not None:
            body["weight"] = weight
        res = await self._t.request("PUT", f"/edges/{edge_id}", json=body)
        return Edge.from_dict(res.data)

    async def delete(self, edge_id: int) -> None:
        await self._t.request("DELETE", f"/edges/{edge_id}")

    async def batch_create(self, edges: Sequence[Mapping[str, Any]]) -> list[Edge]:
        payload = {"edges": [
            {
                "from_node_id": e["from_node_id"],
                "to_node_id": e["to_node_id"],
                "type": e["type"],
                "properties": dict(e.get("properties", {})),
                "weight": float(e.get("weight", 0.0)),
            }
            for e in edges
        ]}
        res = await self._t.request("POST", "/edges/batch", json=payload)
        return [Edge.from_dict(d) for d in (res.data.get("edges") or [])]

    async def list(
        self, *, edge_type: str | None = None, page_size: int = 100
    ) -> AsyncIterator[Edge]:
        """Yield every edge (optionally filtered by type), auto-following X-Next-Cursor."""
        cursor: str | None = None
        prev_cursor: str | None = None
        while True:
            params: dict[str, Any] = {"limit": page_size}
            if edge_type is not None:
                params["type"] = edge_type
            if cursor is not None:
                params["cursor"] = cursor
            res = await self._t.request("GET", "/edges", params=params)
            for d in res.data or []:
                yield Edge.from_dict(d)
            cursor = res.headers.get("X-Next-Cursor")
            if not cursor or cursor == prev_cursor:
                return
            prev_cursor = cursor

File: clients/go/edges.go
package graphdb

import (
	"context"
	"encoding/json"
	"fmt"
	"iter"
	"net/http"
	"net/url"
	"strconv"
)

type EdgeCreateOptions struct {
	Properties map[string]any
	Weight     float64
}

// EdgeUpdateOptions.Weight is a pointer so an omitted weight is not sent
// (the server leaves the edge's weight unchanged rather than zeroing it).
type EdgeUpdateOptions struct {
	Properties map[string]any
	Weight     *float64
}

type EdgeInput struct {
	FromNodeID uint64         `json:"from_node_id"`
	ToNodeID   uint64         `json:"to_node_id"`
	Type       string         `json:"type"`
	Properties map[string]any `json:"properties"`
	Weight     float64        `json:"weight"`
}

func (e *Edges) Create(ctx context.Context, from, to uint64, edgeType string, opts EdgeCreateOptions) (*Edge, error) {
	props := opts.Properties
	if props == nil {
		props = map[string]any{}
	}
	res, err := e.t.request(ctx, http.MethodPost, "/edges", map[string]any{
		"from_node_id": from,
		"to_node_id":   to,
		"type":         edgeType,
		"properties":   props,
		"weight":       opts.Weight,
	}, nil)
	if err != nil {
		return nil, err
	}
	var out Edge
	return &out, json.Unmarshal(res.data, &out)
}

func (e *Edges) Get(ctx context.Context, id uint64) (*Edge, error) {
	res, err := e.t.request(ctx, http.MethodGet, fmt.Sprintf("/edges/%d", id), nil, nil)
	if err != nil {
		return nil, err
	}
	var out Edge
	return &out, json.Unmarshal(res.data, &out)
}

func (e *Edges) Update(ctx context.Context, id uint64, opts EdgeUpdateOptions) (*Edge, error) {
	body := map[string]any{}
	if opts.Properties != nil {
		body["properties"] = opts.Properties
	}
	if opts.Weight != nil {
		body["weight"] = *opts.Weight
	}
	res, err := e.t.request(ctx, http.MethodPut, fmt.Sprintf("/edges/%d", id), body, nil)
	if err != nil {
		return nil, err
	}
	var out Edge
	return &out, json.Unmarshal(res.data, &out)
}

func (e *Edges) Delete(ctx context.Context, id uint64) error {
	_, err := e.t.request(ctx, http.MethodDelete, fmt.Sprintf("/edges/%d", id), nil, nil)
	return err
}

func (e *Edges) BatchCreate(ctx context.Context, edges []EdgeInput) ([]Edge, error) {
	res, err := e.t.request(ctx, http.MethodPost, "/edges/batch",
		map[string]any{"edges": edges}, nil)
	if err != nil {
		return nil, err
	}
	var out struct {
		Edges []Edge `json:"edges"`
	}
	return out.Edges, json.Unmarshal(res.data, &out)
}

// ListEdgesOptions controls edge listing.
type ListEdgesOptions struct {
	Type     string
	PageSize int // default 100 if <= 0
}

// List streams every edge (optionally filtered by type), auto-following the
// X-Next-Cursor response header. On error the iterator yields one final
// (zero, err) pair and stops.
func (e *Edges) List(ctx context.Context, opts ListEdgesOptions) iter.Seq2[Edge, error] {
	pageSize := opts.PageSize
	if pageSize <= 0 {
		pageSize = 100
	}
	return func(yield func(Edge, error) bool) {
		var cursor, prev string
		for {
			params := url.Values{}
			params.Set("limit", strconv.Itoa(pageSize))
			if opts.Type != "" {
				params.Set("type", opts.Type)
			}
			if cursor != "" {
				params.Set("cursor", cursor)
			}
			res, err := e.t.request(ctx, http.MethodGet, "/edges", nil, params)
			if err != nil {
				yield(Edge{}, err)
				return
			}
			var page []Edge
			if err := json.Unmarshal(res.data, &page); err != nil {
				yield(Edge{}, err)
				return
			}
			for _, ed := range page {
				if !yield(ed, nil) {
					return
				}
			}
			cursor = res.header.Get("X-Next-Cursor")
			if cursor == "" || cursor == prev {
				return
			}
			prev = cursor
		}
	}
}

// ListAll collects every edge into a slice (convenience over List).
func (e *Edges) ListAll(ctx context.Context, opts ListEdgesOptions) ([]Edge, error) {
	var out []Edge
	for ed, err := range e.List(ctx, opts) {
		if err != nil {
			return nil, err
		}
		out = append(out, ed)
	}
	return out, nil
}
